crates/driver/src/boundary/liquidity/uniswap/v2.rs

```diff
@@ -0,0 +1,169 @@
+use {
+    crate::{
+        boundary,
+        domain::{eth, liquidity},
+        infra::{
+            self,
+            blockchain::{contracts::ContractAt, Ethereum},
+        },
+    },
+    anyhow::Result,
+    contracts::{GPv2Settlement, IUniswapLikeRouter},
+    futures::StreamExt,
+    shared::{
+        current_block::{self, CurrentBlockStream},
+        maintenance::Maintaining,
+        price_estimation,
+        sources::uniswap_v2::{
+            pair_provider::PairProvider,
+            pool_cache::PoolCache,
+            pool_fetching::{DefaultPoolReader, PoolFetcher, PoolReading},
+        },
+    },
+    solver::{
+        interactions::allowances::{Allowances, NoAllowanceManaging},
+        liquidity::{uniswap_v2, uniswap_v2::UniswapLikeLiquidity, ConstantProductOrder},
+        liquidity_collector::LiquidityCollecting,
+    },
+    std::{
+        sync,
+        sync::{Arc, Mutex},
+    },
+    tracing::Instrument,
+    web3::Web3,
+};
+
+pub fn to_domain(id: liquidity::Id, pool: ConstantProductOrder) -> liquidity::Liquidity {
+    assert!(
+        *pool.fee.numer() == 3 && *pool.fee.denom() == 1000,
+        "uniswap pools have constant fees",
+    );
+
+    let handler = pool
+        .settlement_handling
+        .as_any()
+        .downcast_ref::<uniswap_v2::Inner>()
```

Comment:
I think it is needed because `settlement_handling` is already a trait object. The compiler seems to complain:

With something like:
```rust
let foo: Box<dyn Foo> = Box::new(Bar);
let any: &dyn std::any::Any = &*foo;
```

```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
  --> src/main.rs:15:31
   |
15 | let any: &dyn std::any::Any = &*foo;
   |          ------------------   ^^^^^ expected trait `Any`, found trait `Foo`
   |          |
   |          expected due to this
   |
   = note: expected reference `&(dyn Any + 'static)`
              found reference `&dyn Foo`
```

If I try to make:
```rust
trait Foo: std::any::Any {}
```

Then I get a slightly different error, but still seems not possible:
```
error[[E0658]](https://doc.rust-lang.org/stable/error-index.html#E0658): cannot cast `dyn Foo` to `dyn Any`, trait upcasting coercion is experimental
  --> src/main.rs:15:31
   |
15 | let any: &dyn std::any::Any = &*foo;
   |                               ^^^^^
   |
   = note: [see issue #65991 <https://github.com/rust-lang/rust/issues/65991>](https://github.com/rust-lang/rust/issues/65991) for more information
   = note: required when coercing `&(dyn Foo + 'static)` into `&(dyn Any + 'static)`
```

My quick research online suggests the `as_any` is the workaround for this.